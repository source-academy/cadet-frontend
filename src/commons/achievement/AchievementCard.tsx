import { Icon, Intent, ProgressBar } from '@blueprintjs/core';
import { IconNames } from '@blueprintjs/icons';
import React from 'react';

import { AchievementStatus } from '../../features/achievement/AchievementTypes';
import AchievementDeadline from './utils/AchievementDeadline';
import AchievementExp from './utils/AchievementExp';
import AchievementInferencer from './utils/AchievementInferencer';

type AchievementCardProps = {
  id: number;
  inferencer: AchievementInferencer;
  shouldPartiallyRender: boolean;
  displayView: any;
  handleGlow: any;
  isDropdownOpen?: boolean;
  toggleDropdown?: any;
};

function AchievementCard(props: AchievementCardProps) {
  const {
    id,
    inferencer,
    shouldPartiallyRender,
    displayView,
    handleGlow,
    isDropdownOpen,
    toggleDropdown
  } = props;

  const { title, ability, cardTileUrl } = inferencer.getAchievementItem(id);

  const status = inferencer.getStatus(id);
  const displayDeadline = inferencer.getDisplayDeadline(id);
  const displayExp = inferencer.getMaxExp(id);
  const progressFrac = inferencer.getProgressFrac(id);

  // Only task card with prerequisites has dropdown button
  const hasDropdown =
    isDropdownOpen !== undefined && inferencer.getImmediateChildren(id).size !== 0;

  return (
    <div
      className="achievement-card"
      style={{
        ...handleGlow(id),
        opacity: shouldPartiallyRender ? '20%' : '100%',
        background: `url(${cardTileUrl}) center/cover`
      }}
      onClick={() => displayView(id)}
      onClickCapture={toggleDropdown}
    >
      <div className="dropdown-button">
        {hasDropdown ? (
          <Icon icon={isDropdownOpen ? IconNames.CARET_DOWN : IconNames.CARET_RIGHT} />
        ) : null}
      </div>

      <div className="content">
        <div className="heading">
          <h3>{title.toUpperCase()}</h3>
          <span className="status">
            {status === AchievementStatus.COMPLETED && (
              <Icon icon={IconNames.CONFIRM} intent={Intent.SUCCESS} style={{ padding: '1em' }} />
            )}
          </span>
        </div>

        <div className="details">
          <div className="ability">
            <p>{ability}</p>
          </div>
          <AchievementDeadline deadline={displayDeadline} ability={ability} />
          <AchievementExp exp={displayExp} isBonus={hasDropdown} />
        </div>

        <ProgressBar
          className="progress"
          intent={progressFrac === 1 ? 'success' : undefined}
          value={progressFrac}
          animate={false}
          stripes={false}
        />
      </div>
    </div>
  );
}

export default AchievementCard;
