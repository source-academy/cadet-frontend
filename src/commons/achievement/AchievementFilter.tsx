import { Icon, IconName } from '@blueprintjs/core';
import React from 'react';

import { FilterStatus } from '../../features/achievement/AchievementTypes';

type AchievementFilterProps = {
  filterStatus: FilterStatus;
  setFilterStatus: any;
  icon: IconName;
  count: number;
  handleFilterColor: any;
};

function AchievementFilter(props: AchievementFilterProps) {
  const { filterStatus, setFilterStatus, icon, count, handleFilterColor } = props;

  /**
   * Changes the filter status for the achievement page.
   */
  const changeFilterStatus = () => {
    setFilterStatus(filterStatus);
  };

  const filterColor = handleFilterColor(filterStatus);

  return (
    <div className="filter-button" onClick={changeFilterStatus}>
      <Icon iconSize={40} icon={icon} color={filterColor} />
      <br />
      <span style={{ color: filterColor }}>
        {filterStatus} [{count}]
      </span>
    </div>
  );
}

export default AchievementFilter;
