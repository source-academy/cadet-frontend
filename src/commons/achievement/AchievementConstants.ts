import { Links } from '../utils/Constants';
import { AchievementAbility } from './AchievementTypes';

export const expPerLevel = 1000;

export const abilityColor = (ability: AchievementAbility) => {
  switch (ability) {
    case AchievementAbility.CORE:
      return '#ffb412';
    case AchievementAbility.EFFORT:
      return '#b5ff61';
    case AchievementAbility.EXPLORATION:
      return '#9ecaed';
    case AchievementAbility.COMMUNITY:
      return '#ff6780';
    default:
      return '#ffffff';
  }
};

export const backgroundUrl = `${Links.sourceAcademyAssets}/achievement/background`;

export const abilityBackground = (ability: AchievementAbility) => {
  switch (ability) {
    case AchievementAbility.CORE:
      return {
        background: `url(${backgroundUrl}/core-background.png) no-repeat center`
      };
    case AchievementAbility.EFFORT:
      return {
        background: `url(${backgroundUrl}/effort-background.png) no-repeat center`
      };
    case AchievementAbility.EXPLORATION:
      return {
        background: `url(${backgroundUrl}/exploration-background.png) no-repeat center`
      };
    case AchievementAbility.COMMUNITY:
      return {
        background: `url(${backgroundUrl}/community-background.png) no-repeat center`
      };
    default:
      return {
        background: 'black'
      };
  }
};
