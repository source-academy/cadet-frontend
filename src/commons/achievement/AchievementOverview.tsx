import React, { useContext } from 'react';
import { AchievementContext } from 'src/features/achievement/AchievementConstants';

import AchievementLevel from './overview/AchievementLevel';

type AchievementOverviewProps = {
  name: string;
  studio: string;
};

function AchievementOverview(props: AchievementOverviewProps) {
  const { name, studio } = props;

  const inferencer = useContext(AchievementContext);
  const studentXp = inferencer.getTotalXp();

  return (
    <div className="achievement-overview">
      <AchievementLevel studentXp={studentXp} />
      <h3>{name}</h3>
      <h3>{studio}</h3>
    </div>
  );
}

export default AchievementOverview;
