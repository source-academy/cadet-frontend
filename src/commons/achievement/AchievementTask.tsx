import React, { useState } from 'react';

import { getAbilityColor } from '../../features/achievement/AchievementConstants';
import { AchievementStatus, FilterStatus } from '../../features/achievement/AchievementTypes';
import AchievementCard from './AchievementCard';
import AchievementInferencer from './utils/AchievementInferencer';

type AchievementTaskProps = {
  id: number;
  inferencer: AchievementInferencer;
  filterStatus: FilterStatus;
  displayView: any;
  handleGlow: any;
};

function AchievementTask(props: AchievementTaskProps) {
  const { id, inferencer, filterStatus, displayView, handleGlow } = props;

  const [isDropdownOpen, setIsDropdownOpen] = useState<boolean>(false);
  const toggleDropdown = () => {
    setIsDropdownOpen(!isDropdownOpen);
  };

  const taskColor = getAbilityColor(inferencer.getAchievementItem(id).ability);

  /**
   * Checks whether the AchievementItem (can be a task or prereq) should be rendered
   * based on the achivement page filterStatus.
   */
  const shouldRender = (id: number): boolean => {
    const status = inferencer.getStatus(id);
    switch (filterStatus) {
      case FilterStatus.ALL:
        return true;
      case FilterStatus.ACTIVE:
        return status === AchievementStatus.ACTIVE;
      case FilterStatus.COMPLETED:
        return status === AchievementStatus.COMPLETED;
      default:
        return false;
    }
  };

  /**
   * Checks whether the achievement item has any prerequisite item that
   * should be rendered based on the achievement page filterStatus.
   *
   * If there is at least 1 prerequisite that needs to be rendered,
   * the whole AchievementTask will be rendered together.
   */
  const shouldRenderPrerequisites = (id: number) => {
    const children = inferencer.listImmediateChildren(id);
    return children.reduce((canRender, prerequisite) => {
      return canRender || shouldRender(prerequisite);
    }, false);
  };

  /**
   * If the main achievement or any of the prerequisites need to be rendered,
   * the whole achievement task will be rendered
   */
  const shouldRenderTask = (id: number) => shouldRender(id) || shouldRenderPrerequisites(id);

  return (
    <>
      {shouldRenderTask(id) && (
        <li key={id}>
          <AchievementCard
            id={id}
            inferencer={inferencer}
            shouldPartiallyRender={!shouldRender(id)}
            displayView={displayView}
            handleGlow={handleGlow}
            isDropdownOpen={isDropdownOpen}
            toggleDropdown={toggleDropdown}
          />
          {isDropdownOpen && (
            <div className="prerequisite-container">
              {inferencer.listImmediateChildren(id).map(prerequisite => (
                <div className="prerequisite" key={prerequisite}>
                  <div
                    className="dropdown-lines"
                    style={{
                      borderBottom: `1px solid ${taskColor}`,
                      borderLeft: `1px solid ${taskColor}`
                    }}
                  ></div>
                  <AchievementCard
                    id={prerequisite}
                    inferencer={inferencer}
                    shouldPartiallyRender={!shouldRender(prerequisite)}
                    displayView={displayView}
                    handleGlow={handleGlow}
                  />
                </div>
              ))}
            </div>
          )}
        </li>
      )}
    </>
  );
}

export default AchievementTask;
