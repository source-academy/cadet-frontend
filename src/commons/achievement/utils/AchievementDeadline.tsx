import { Icon } from '@blueprintjs/core';
import { IconNames } from '@blueprintjs/icons';

import { DeadlineColors } from '../../../features/achievement/AchievementConstants';
import { AchievementAbility } from '../../../features/achievement/AchievementTypes';
import { isExpired, prettifyDeadline, timeFromExpired } from './DateHelper';

type AchievementDeadlineProps = {
  deadline?: Date;
  ability: AchievementAbility;
};

const twoDays = new Date(0, 0, 2).getTime() - new Date(0, 0, 0).getTime();

function AchievementDeadline(props: AchievementDeadlineProps) {
  const { deadline, ability } = props;

  // red deadline color for core achievements that are expiring in less than 2 days
  const deadlineColor =
    ability === AchievementAbility.CORE &&
    deadline !== undefined &&
    !isExpired(deadline) &&
    timeFromExpired(deadline) <= twoDays
      ? DeadlineColors.RED
      : DeadlineColors.BLACK;

  return (
    <div className="deadline">
      <Icon icon={IconNames.STOPWATCH} color={deadlineColor} />
      <span style={{ color: deadlineColor }}>
        <p>{prettifyDeadline(deadline)}</p>
      </span>
    </div>
  );
}

export default AchievementDeadline;
