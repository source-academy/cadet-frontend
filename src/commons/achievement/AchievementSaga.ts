import { SagaIterator } from 'redux-saga';
import { call, put, select, takeEvery } from 'redux-saga/effects';

import { OverallState } from '../application/ApplicationTypes';
import {
  editAchievement,
  getAchievements,
  removeAchievement,
  removeGoal
} from '../sagas/RequestsSaga';
import { actions } from '../utils/ActionsHelper';
import {
  EDIT_ACHIEVEMENT,
  GET_ACHIEVEMENTS,
  REMOVE_ACHIEVEMENT,
  REMOVE_GOAL
} from './AchievementTypes';

export default function* AchievementSaga(): SagaIterator {
  yield takeEvery(GET_ACHIEVEMENTS, function* () {
    const tokens = yield select((state: OverallState) => ({
      accessToken: state.session.accessToken,
      refreshToken: state.session.refreshToken
    }));

    const achievements = yield call(getAchievements, tokens);

    if (achievements) {
      yield put(actions.saveAchievements(achievements));
    }
  });

  yield takeEvery(EDIT_ACHIEVEMENT, function* (action: ReturnType<typeof actions.editAchievement>) {
    const tokens = yield select((state: OverallState) => ({
      accessToken: state.session.accessToken,
      refreshToken: state.session.refreshToken
    }));

    const achievement = action.payload;

    const resp = yield call(editAchievement, achievement, tokens);

    if (!resp) {
      return;
    }
  });

  yield takeEvery(REMOVE_ACHIEVEMENT, function* (
    action: ReturnType<typeof actions.removeAchievement>
  ) {
    const tokens = yield select((state: OverallState) => ({
      accessToken: state.session.accessToken,
      refreshToken: state.session.refreshToken
    }));

    const achievement = action.payload;

    const resp = yield call(removeAchievement, achievement, tokens);

    if (!resp) {
      return;
    }
  });

  yield takeEvery(REMOVE_GOAL, function* (action: ReturnType<typeof actions.removeGoal>) {
    const tokens = yield select((state: OverallState) => ({
      accessToken: state.session.accessToken,
      refreshToken: state.session.refreshToken
    }));

    const { goal, achievement } = action.payload;

    const resp = yield call(removeGoal, goal, achievement, tokens);

    if (!resp) {
      return;
    }
  });
}
