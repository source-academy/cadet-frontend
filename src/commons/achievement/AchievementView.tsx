import React from 'react';

import {
  getAbilityBackground,
  getAbilityGlow
} from '../../features/achievement/AchievementConstants';
import { AchievementStatus } from '../../features/achievement/AchievementTypes';
import AchievementInferencer from './utils/AchievementInferencer';
import { prettifyDate } from './utils/DateHelper';
import AchievementViewCompletion from './view/AchievementViewCompletion';
import AchievementViewGoal from './view/AchievementViewGoal';

type AchievementViewProps = {
  inferencer: AchievementInferencer;
  focusId: number;
};

function AchievementView(props: AchievementViewProps) {
  const { inferencer, focusId } = props;

  if (focusId < 0) return null;

  const achievement = inferencer.getAchievementItem(focusId);
  const { ability, deadline, title, view } = achievement;
  const { canvasUrl, completionText, description } = view;

  const awardedExp = inferencer.getExp(focusId);
  const goals = inferencer.getGoals(focusId);
  const prereqGoals = inferencer.getPrerequisiteGoals(focusId);
  const status = inferencer.getStatus(focusId);

  return (
    <div className="view" style={{ ...getAbilityGlow(ability), ...getAbilityBackground(ability) }}>
      <div
        className="canvas"
        style={{
          background: `url(${canvasUrl}) center/cover`
        }}
      >
        <h1>{title.toUpperCase()}</h1>
        {deadline && <p>{`Deadline: ${prettifyDate(deadline)}`}</p>}
        <span className="description">
          <p>{description}</p>
        </span>
      </div>
      <AchievementViewGoal goals={goals} />
      {prereqGoals.length > 0 && (
        <>
          <hr />
          <AchievementViewGoal goals={prereqGoals} />
        </>
      )}
      {status === AchievementStatus.COMPLETED && (
        <>
          <hr />
          <AchievementViewCompletion awardedExp={awardedExp} completionText={completionText} />
        </>
      )}
    </div>
  );
}

export default AchievementView;
