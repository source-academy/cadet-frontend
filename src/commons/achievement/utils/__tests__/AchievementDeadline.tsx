import { mount } from 'enzyme';
import React from 'react';

import { AchievementAbility } from '../../../../features/achievement/AchievementTypes';
import AchievementDeadline from '../AchievementDeadline';

const undefinedDeadlineProps = {
  deadline: undefined,
  ability: AchievementAbility.CORE
};

const expiredDeadlineProps = {
  deadline: new Date(),
  ability: AchievementAbility.CORE
};

const urgentDeadlineProps = {
  deadline: new Date(Date.now() + 1000 * 60 * 60 * 2),
  ability: AchievementAbility.CORE
};

const daysDeadlineProps = {
  deadline: new Date(Date.now() + 1000 * 60 * 60 * 25),
  ability: AchievementAbility.CORE
};

const weeksDeadlineProps = {
  deadline: new Date(Date.now() + 1000 * 60 * 60 * 200),
  ability: AchievementAbility.CORE
};

test('AchievementDeadline component renders correctly', () => {
  const expiredDeadline = <AchievementDeadline {...expiredDeadlineProps} />;
  const tree_1 = mount(expiredDeadline);
  expect(tree_1.debug()).toMatchSnapshot();

  const urgentDeadline = <AchievementDeadline {...urgentDeadlineProps} />;
  const tree_2 = mount(urgentDeadline);
  expect(tree_2.debug()).toMatchSnapshot();

  const daysDeadline = <AchievementDeadline {...daysDeadlineProps} />;
  const tree_3 = mount(daysDeadline);
  expect(tree_3.debug()).toMatchSnapshot();

  const weeksDeadline = <AchievementDeadline {...weeksDeadlineProps} />;
  const tree_4 = mount(weeksDeadline);
  expect(tree_4.debug()).toMatchSnapshot();

  const undefinedDeadline = <AchievementDeadline {...undefinedDeadlineProps} />;
  const tree = mount(undefinedDeadline);
  expect(tree.debug()).toMatchSnapshot();
});
