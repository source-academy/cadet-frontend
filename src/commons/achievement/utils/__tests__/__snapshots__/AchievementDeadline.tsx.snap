// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AchievementDeadline component renders correctly 1`] = `
"<AchievementDeadline deadline={{...}} ability=\\"Core\\">
  <div className=\\"deadline\\">
    <Blueprint3.Icon icon=\\"stopwatch\\" color=\\"#000000\\">
      <span icon=\\"stopwatch\\" className=\\"bp3-icon bp3-icon-stopwatch\\" title={[undefined]}>
        <svg fill=\\"#000000\\" data-icon=\\"stopwatch\\" width={16} height={16} viewBox=\\"0 0 16 16\\">
          <desc>
            stopwatch
          </desc>
          <path d=\\"M9 2v1.083A6.002 6.002 0 018 15 6 6 0 017 3.083V2H6a1 1 0 110-2h4a1 1 0 010 2H9zM8 5a4 4 0 104 4H8V5z\\" fillRule=\\"evenodd\\" />
        </svg>
      </span>
    </Blueprint3.Icon>
    <span style={{...}}>
      <p>
        Expired
      </p>
    </span>
  </div>
</AchievementDeadline>"
`;

exports[`AchievementDeadline component renders correctly 2`] = `
"<AchievementDeadline deadline={{...}} ability=\\"Core\\">
  <div className=\\"deadline\\">
    <Blueprint3.Icon icon=\\"stopwatch\\" color=\\"#ff0000\\">
      <span icon=\\"stopwatch\\" className=\\"bp3-icon bp3-icon-stopwatch\\" title={[undefined]}>
        <svg fill=\\"#ff0000\\" data-icon=\\"stopwatch\\" width={16} height={16} viewBox=\\"0 0 16 16\\">
          <desc>
            stopwatch
          </desc>
          <path d=\\"M9 2v1.083A6.002 6.002 0 018 15 6 6 0 017 3.083V2H6a1 1 0 110-2h4a1 1 0 010 2H9zM8 5a4 4 0 104 4H8V5z\\" fillRule=\\"evenodd\\" />
        </svg>
      </span>
    </Blueprint3.Icon>
    <span style={{...}}>
      <p>
        2 Hours
      </p>
    </span>
  </div>
</AchievementDeadline>"
`;

exports[`AchievementDeadline component renders correctly 3`] = `
"<AchievementDeadline deadline={{...}} ability=\\"Core\\">
  <div className=\\"deadline\\">
    <Blueprint3.Icon icon=\\"stopwatch\\" color=\\"#ff0000\\">
      <span icon=\\"stopwatch\\" className=\\"bp3-icon bp3-icon-stopwatch\\" title={[undefined]}>
        <svg fill=\\"#ff0000\\" data-icon=\\"stopwatch\\" width={16} height={16} viewBox=\\"0 0 16 16\\">
          <desc>
            stopwatch
          </desc>
          <path d=\\"M9 2v1.083A6.002 6.002 0 018 15 6 6 0 017 3.083V2H6a1 1 0 110-2h4a1 1 0 010 2H9zM8 5a4 4 0 104 4H8V5z\\" fillRule=\\"evenodd\\" />
        </svg>
      </span>
    </Blueprint3.Icon>
    <span style={{...}}>
      <p>
        2 Days
      </p>
    </span>
  </div>
</AchievementDeadline>"
`;

exports[`AchievementDeadline component renders correctly 4`] = `
"<AchievementDeadline deadline={{...}} ability=\\"Core\\">
  <div className=\\"deadline\\">
    <Blueprint3.Icon icon=\\"stopwatch\\" color=\\"#000000\\">
      <span icon=\\"stopwatch\\" className=\\"bp3-icon bp3-icon-stopwatch\\" title={[undefined]}>
        <svg fill=\\"#000000\\" data-icon=\\"stopwatch\\" width={16} height={16} viewBox=\\"0 0 16 16\\">
          <desc>
            stopwatch
          </desc>
          <path d=\\"M9 2v1.083A6.002 6.002 0 018 15 6 6 0 017 3.083V2H6a1 1 0 110-2h4a1 1 0 010 2H9zM8 5a4 4 0 104 4H8V5z\\" fillRule=\\"evenodd\\" />
        </svg>
      </span>
    </Blueprint3.Icon>
    <span style={{...}}>
      <p>
        2 Weeks
      </p>
    </span>
  </div>
</AchievementDeadline>"
`;

exports[`AchievementDeadline component renders correctly 5`] = `
"<AchievementDeadline deadline={[undefined]} ability=\\"Core\\">
  <div className=\\"deadline\\">
    <Blueprint3.Icon icon=\\"stopwatch\\" color=\\"#000000\\">
      <span icon=\\"stopwatch\\" className=\\"bp3-icon bp3-icon-stopwatch\\" title={[undefined]}>
        <svg fill=\\"#000000\\" data-icon=\\"stopwatch\\" width={16} height={16} viewBox=\\"0 0 16 16\\">
          <desc>
            stopwatch
          </desc>
          <path d=\\"M9 2v1.083A6.002 6.002 0 018 15 6 6 0 017 3.083V2H6a1 1 0 110-2h4a1 1 0 010 2H9zM8 5a4 4 0 104 4H8V5z\\" fillRule=\\"evenodd\\" />
        </svg>
      </span>
    </Blueprint3.Icon>
    <span style={{...}}>
      <p>
        Unlimited
      </p>
    </span>
  </div>
</AchievementDeadline>"
`;
